extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProvider.java

```diff
@@ -119,56 +128,87 @@ public SeekableByteChannel getWriteChannel(@NotNull final Path path, final boole
+            }
+
+            private void fetchNextBatch() throws IOException {
+                currentIndex = 0;
+                currentBatch.clear();
+                final S3Uri s3DirectoryURI = s3AsyncClient.utilities().parseUri(directory);
+                final String bucketName = s3DirectoryURI.bucket().orElseThrow();
+                final String directoryKey = s3DirectoryURI.key().orElseThrow();
+                final ListObjectsV2Request.Builder requestBuilder = ListObjectsV2Request.builder()
+                        .bucket(bucketName)
+                        .prefix(s3DirectoryURI.key().orElseThrow());
+                if (!isRecursive) {
+                    // Add a delimiter to the request if we don't want to fetch all files recursively
+                    requestBuilder.delimiter("/");
+                }
+                final long readTimeoutNanos = s3Instructions.readTimeout().toNanos();
+                final ListObjectsV2Response response;
+                final ListObjectsV2Request request = requestBuilder.continuationToken(continuationToken).build();
+                try {
+                    response = s3AsyncClient.listObjectsV2(request).get(readTimeoutNanos, TimeUnit.NANOSECONDS);
+                } catch (final InterruptedException | ExecutionException | TimeoutException | CancellationException e) {
+                    throw handleS3Exception(e, String.format("fetching list of files in directory %s", directory),
+                            s3Instructions);
+                }
+                response.contents().stream()
+                        .filter(s3Object -> !s3Object.key().equals(directoryKey))
+                        .map(s3Object -> URI.create("s3://" + bucketName + "/" + s3Object.key()))
+                        .forEach(currentBatch::add);
+                // If the response is truncated, fetch the next page using the continuation token from the response
+                // Usually the response is truncated when there are more than 1000 objects in the bucket
+                // TODO Test this
+                fetchedAllURIs = !response.isTruncated();
+                continuationToken = response.nextContinuationToken();
             }
-            response.contents().stream()
-                    .filter(s3Object -> !s3Object.key().equals(directoryKey)) // Skip the directory itself
-                    .map(s3Object -> URI.create("s3://" + bucketName + "/" + s3Object.key()))
-                    .forEach(processor);
-
-            // If the response is truncated, fetch the next page using the continuation token from the response
-            // Usually the response is truncated when there are more than 1000 objects in the bucket
-            // TODO Test this
-            continuationToken = response.nextContinuationToken();
-        } while (response.isTruncated());
+        };
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, Spliterator.ORDERED), false);
```

Comment:
Not sure how much we technically care about finely specifying the characteristics of the stream since our downstream consumers are simple... but at a minimum we could set `DISTINCT`, `NONNULL`. _Maybe_ also be `IMMUTABLE`, but I don't fully grok the semantics.

I _assumed_ that it would also be `SORTED`. Looking at https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html, I learned:

Sorting order of returned objects
* General purpose bucket - For general purpose buckets, ListObjectsV2 returns objects in lexicographical order based on their key names.
* Directory bucket - For directory buckets, ListObjectsV2 does not return objects in lexicographical order.

I'd never heard of "directory bucket" https://docs.aws.amazon.com/AmazonS3/latest/userguide/directory-buckets-overview.html

* Directory buckets use the S3 Express One Zone storage class, which is recommended if your application is performance sensitive and benefits from single-digit millisecond PUT and GET latencies.

I wonder if we should be doing performance testing w/ directory buckets as well?